// @snap/ts-inject/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PartialContainer = exports.ConcatInjectable = exports.InjectableCompat = exports.Injectable = exports.Container = exports.CONTAINER = void 0;
var Container_1 = require("./Container");
Object.defineProperty(exports, "CONTAINER", { enumerable: true, get: function () { return Container_1.CONTAINER; } });
Object.defineProperty(exports, "Container", { enumerable: true, get: function () { return Container_1.Container; } });
var Injectable_1 = require("./Injectable");
Object.defineProperty(exports, "Injectable", { enumerable: true, get: function () { return Injectable_1.Injectable; } });
Object.defineProperty(exports, "InjectableCompat", { enumerable: true, get: function () { return Injectable_1.InjectableCompat; } });
Object.defineProperty(exports, "ConcatInjectable", { enumerable: true, get: function () { return Injectable_1.ConcatInjectable; } });
var PartialContainer_1 = require("./PartialContainer");
Object.defineProperty(exports, "PartialContainer", { enumerable: true, get: function () { return PartialContainer_1.PartialContainer; } });
