// @snap/camera-kit/dist/generated-proto/pb_schema/camera_kit/v3/features/ranking_info.js
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
export const protobufPackage = "com.snap.camerakit.v3.features";
function createBaseRankingInfo() {
    return { rankingRequestId: "" };
}
export const RankingInfo = {
    encode(message, writer = new BinaryWriter()) {
        if (message.rankingRequestId !== "") {
            writer.uint32(10).string(message.rankingRequestId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRankingInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.rankingRequestId = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    create(base) {
        return RankingInfo.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseRankingInfo();
        message.rankingRequestId = (_a = object.rankingRequestId) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
