// @snap/camera-kit/dist/generated-proto/pb_schema/content/asset.js
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
export const protobufPackage = "snapchat.content";
export var Asset_AssetType;
(function (Asset_AssetType) {
    Asset_AssetType[Asset_AssetType["UNSET"] = 0] = "UNSET";
    Asset_AssetType[Asset_AssetType["BLOOPS_STICKER"] = 1] = "BLOOPS_STICKER";
    Asset_AssetType[Asset_AssetType["MUSIC"] = 2] = "MUSIC";
    Asset_AssetType[Asset_AssetType["RAW_MEDIA"] = 3] = "RAW_MEDIA";
    Asset_AssetType[Asset_AssetType["CANVAS_STICKER"] = 4] = "CANVAS_STICKER";
    Asset_AssetType[Asset_AssetType["BASE_MEDIA"] = 5] = "BASE_MEDIA";
    Asset_AssetType[Asset_AssetType["OVERLAY"] = 6] = "OVERLAY";
    Asset_AssetType[Asset_AssetType["EDITS"] = 7] = "EDITS";
    Asset_AssetType[Asset_AssetType["FIRST_FRAME"] = 8] = "FIRST_FRAME";
    Asset_AssetType[Asset_AssetType["THUMBNAIL"] = 9] = "THUMBNAIL";
    Asset_AssetType[Asset_AssetType["CAMEO"] = 10] = "CAMEO";
    Asset_AssetType[Asset_AssetType["WEB_THUMBNAIL"] = 11] = "WEB_THUMBNAIL";
    Asset_AssetType[Asset_AssetType["WEB_MEDIA"] = 12] = "WEB_MEDIA";
    Asset_AssetType[Asset_AssetType["LENS_ASSET"] = 13] = "LENS_ASSET";
    Asset_AssetType[Asset_AssetType["VOICEOVER"] = 14] = "VOICEOVER";
    Asset_AssetType[Asset_AssetType["AVATAR_IMAGE"] = 15] = "AVATAR_IMAGE";
    Asset_AssetType[Asset_AssetType["SPECTACLES_SIXDOF"] = 16] = "SPECTACLES_SIXDOF";
    Asset_AssetType[Asset_AssetType["PLAIN_MEDIA"] = 17] = "PLAIN_MEDIA";
    Asset_AssetType[Asset_AssetType["LENS_MEDIA"] = 18] = "LENS_MEDIA";
    Asset_AssetType[Asset_AssetType["LOCAL_BURN_IN_ASSET"] = 19] = "LOCAL_BURN_IN_ASSET";
    Asset_AssetType[Asset_AssetType["CAMERA_ROLL_INDEX"] = 20] = "CAMERA_ROLL_INDEX";
    Asset_AssetType[Asset_AssetType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(Asset_AssetType || (Asset_AssetType = {}));
function createBaseAsset() {
    return { id: "", type: 0 };
}
export const Asset = {
    encode(message, writer = new BinaryWriter()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.type !== 0) {
            writer.uint32(16).int32(message.type);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAsset();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.type = reader.int32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    create(base) {
        return Asset.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseAsset();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "";
        message.type = (_b = object.type) !== null && _b !== void 0 ? _b : 0;
        return message;
    },
};
