// @snap/camera-kit/dist/generated-proto/pb_schema/content/media_source.js
export const protobufPackage = "snapchat.content.attribution";
export var MediaSource;
(function (MediaSource) {
    MediaSource[MediaSource["UNSET"] = 0] = "UNSET";
    MediaSource[MediaSource["CHAT"] = 1] = "CHAT";
    MediaSource[MediaSource["MEMORIES"] = 2] = "MEMORIES";
    MediaSource[MediaSource["LENS"] = 3] = "LENS";
    MediaSource[MediaSource["MAPS"] = 4] = "MAPS";
    MediaSource[MediaSource["MUSIC"] = 5] = "MUSIC";
    MediaSource[MediaSource["CAMEOS"] = 6] = "CAMEOS";
    MediaSource[MediaSource["COMMERCE"] = 7] = "COMMERCE";
    MediaSource[MediaSource["STICKERS"] = 8] = "STICKERS";
    MediaSource[MediaSource["PROFILE"] = 9] = "PROFILE";
    MediaSource[MediaSource["SNAP_PRO"] = 10] = "SNAP_PRO";
    MediaSource[MediaSource["EXPORTED_MEDIA_SHARE"] = 11] = "EXPORTED_MEDIA_SHARE";
    MediaSource[MediaSource["FORMA"] = 12] = "FORMA";
    MediaSource[MediaSource["AD_WEBVIEW_HTML"] = 13] = "AD_WEBVIEW_HTML";
    MediaSource[MediaSource["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(MediaSource || (MediaSource = {}));
