// @snap/camera-kit/dist/generated-proto/pb_schema/content/media_types.js
export const protobufPackage = "snapchat.content";
export var MediaType;
(function (MediaType) {
    MediaType[MediaType["Unassigned"] = 0] = "Unassigned";
    MediaType[MediaType["Overlay"] = 1] = "Overlay";
    MediaType[MediaType["Image"] = 2] = "Image";
    MediaType[MediaType["Video"] = 3] = "Video";
    MediaType[MediaType["AssetBundle"] = 4] = "AssetBundle";
    MediaType[MediaType["Audio"] = 5] = "Audio";
    MediaType[MediaType["AnimatedImage"] = 6] = "AnimatedImage";
    MediaType[MediaType["Font"] = 7] = "Font";
    MediaType[MediaType["WebViewContent"] = 8] = "WebViewContent";
    MediaType[MediaType["VideoNoAudio"] = 9] = "VideoNoAudio";
    MediaType[MediaType["Subtitles"] = 10] = "Subtitles";
    MediaType[MediaType["DataBlob"] = 11] = "DataBlob";
    MediaType[MediaType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(MediaType || (MediaType = {}));
export var MediaContextType;
(function (MediaContextType) {
    MediaContextType[MediaContextType["Bitmoji"] = 0] = "Bitmoji";
    MediaContextType[MediaContextType["LensImages"] = 1] = "LensImages";
    MediaContextType[MediaContextType["Stickers"] = 2] = "Stickers";
    MediaContextType[MediaContextType["ChatSnap"] = 3] = "ChatSnap";
    MediaContextType[MediaContextType["FriendStorySnap"] = 4] = "FriendStorySnap";
    MediaContextType[MediaContextType["TempUnassigned"] = 5] = "TempUnassigned";
    MediaContextType[MediaContextType["FriendStoryThumbnail"] = 6] = "FriendStoryThumbnail";
    MediaContextType[MediaContextType["RecommendedStoryThumbnail"] = 7] = "RecommendedStoryThumbnail";
    MediaContextType[MediaContextType["CognacImages"] = 8] = "CognacImages";
    MediaContextType[MediaContextType["CommerceImages"] = 9] = "CommerceImages";
    MediaContextType[MediaContextType["Ads"] = 10] = "Ads";
    MediaContextType[MediaContextType["ComposerImages"] = 11] = "ComposerImages";
    MediaContextType[MediaContextType["Lens"] = 12] = "Lens";
    MediaContextType[MediaContextType["OnDemand"] = 13] = "OnDemand";
    MediaContextType[MediaContextType["Onboarding"] = 14] = "Onboarding";
    MediaContextType[MediaContextType["Bloops"] = 15] = "Bloops";
    MediaContextType[MediaContextType["RecommendedUserStorySnap"] = 16] = "RecommendedUserStorySnap";
    MediaContextType[MediaContextType["LocalizationStrings"] = 17] = "LocalizationStrings";
    MediaContextType[MediaContextType["Music"] = 18] = "Music";
    MediaContextType[MediaContextType["MemoriesSnap"] = 19] = "MemoriesSnap";
    MediaContextType[MediaContextType["PublisherStorySnap"] = 20] = "PublisherStorySnap";
    MediaContextType[MediaContextType["GiphyStickers"] = 21] = "GiphyStickers";
    MediaContextType[MediaContextType["UGCLensAssets"] = 22] = "UGCLensAssets";
    MediaContextType[MediaContextType["FriendStoryFirstFrame"] = 23] = "FriendStoryFirstFrame";
    MediaContextType[MediaContextType["RecommendedUserStoryFirstFrame"] = 24] = "RecommendedUserStoryFirstFrame";
    MediaContextType[MediaContextType["Geofilter"] = 25] = "Geofilter";
    MediaContextType[MediaContextType["MemoriesThumbnail"] = 26] = "MemoriesThumbnail";
    MediaContextType[MediaContextType["LensBitmoji"] = 27] = "LensBitmoji";
    MediaContextType[MediaContextType["PublisherStoryFirstFrame"] = 28] = "PublisherStoryFirstFrame";
    MediaContextType[MediaContextType["Maps"] = 29] = "Maps";
    MediaContextType[MediaContextType["Snapshots"] = 30] = "Snapshots";
    MediaContextType[MediaContextType["SystemTestImages"] = 31] = "SystemTestImages";
    MediaContextType[MediaContextType["SystemTestVideos"] = 32] = "SystemTestVideos";
    MediaContextType[MediaContextType["ARBusiness"] = 33] = "ARBusiness";
    MediaContextType[MediaContextType["PreviewSnap"] = 34] = "PreviewSnap";
    MediaContextType[MediaContextType["Spectacles"] = 35] = "Spectacles";
    MediaContextType[MediaContextType["Cheerios"] = 36] = "Cheerios";
    MediaContextType[MediaContextType["ExportedMediaShare"] = 37] = "ExportedMediaShare";
    MediaContextType[MediaContextType["InclusionPanelSurvey"] = 38] = "InclusionPanelSurvey";
    MediaContextType[MediaContextType["Communities"] = 39] = "Communities";
    MediaContextType[MediaContextType["SpotlightSnap"] = 40] = "SpotlightSnap";
    MediaContextType[MediaContextType["TosHtml"] = 41] = "TosHtml";
    MediaContextType[MediaContextType["ComposerModulesArchive"] = 42] = "ComposerModulesArchive";
    MediaContextType[MediaContextType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(MediaContextType || (MediaContextType = {}));
