// @snap/camera-kit/dist/generated-proto/pb_schema/cdp/cof/content_manager_provided_signals.js
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
export const protobufPackage = "snapchat.cdp.cof";
export var LegacyMushroomContentType;
(function (LegacyMushroomContentType) {
    LegacyMushroomContentType[LegacyMushroomContentType["UNKNOWN_ASSET_TYPE"] = 0] = "UNKNOWN_ASSET_TYPE";
    LegacyMushroomContentType[LegacyMushroomContentType["GENERIC_LARGE_ASSET_TYPE"] = 1] = "GENERIC_LARGE_ASSET_TYPE";
    LegacyMushroomContentType[LegacyMushroomContentType["GENERIC_SMALL_ASSET_TYPE"] = 2] = "GENERIC_SMALL_ASSET_TYPE";
    LegacyMushroomContentType[LegacyMushroomContentType["AD_REMOTE_ASSET"] = 3] = "AD_REMOTE_ASSET";
    LegacyMushroomContentType[LegacyMushroomContentType["AD_SNAP"] = 4] = "AD_SNAP";
    LegacyMushroomContentType[LegacyMushroomContentType["AD_WEB_VIEW_RESOURCE_CONTENT"] = 5] = "AD_WEB_VIEW_RESOURCE_CONTENT";
    LegacyMushroomContentType[LegacyMushroomContentType["ATTACHMENT_INFO"] = 6] = "ATTACHMENT_INFO";
    LegacyMushroomContentType[LegacyMushroomContentType["BITMOJI"] = 7] = "BITMOJI";
    LegacyMushroomContentType[LegacyMushroomContentType["BITMOJI_LENS_AVATAR_ASSET"] = 8] = "BITMOJI_LENS_AVATAR_ASSET";
    LegacyMushroomContentType[LegacyMushroomContentType["BITMOJI_LENS_METADATA"] = 9] = "BITMOJI_LENS_METADATA";
    LegacyMushroomContentType[LegacyMushroomContentType["CAPTION_METADATA"] = 10] = "CAPTION_METADATA";
    LegacyMushroomContentType[LegacyMushroomContentType["CAPTION_TYPEFACE"] = 11] = "CAPTION_TYPEFACE";
    LegacyMushroomContentType[LegacyMushroomContentType["CAPTION_VIEW_BITMAP"] = 12] = "CAPTION_VIEW_BITMAP";
    LegacyMushroomContentType[LegacyMushroomContentType["CHAT_MEDIA_THUMBNAIL"] = 13] = "CHAT_MEDIA_THUMBNAIL";
    LegacyMushroomContentType[LegacyMushroomContentType["COGNAC"] = 14] = "COGNAC";
    LegacyMushroomContentType[LegacyMushroomContentType["COGNAC_WEBVIEW"] = 15] = "COGNAC_WEBVIEW";
    LegacyMushroomContentType[LegacyMushroomContentType["COMMERCE"] = 16] = "COMMERCE";
    LegacyMushroomContentType[LegacyMushroomContentType["CONTEXT_FILTER_METADATA"] = 17] = "CONTEXT_FILTER_METADATA";
    LegacyMushroomContentType[LegacyMushroomContentType["CUSTOM_STICKERS"] = 18] = "CUSTOM_STICKERS";
    LegacyMushroomContentType[LegacyMushroomContentType["DISCOVER_FEED_THUMBNAIL"] = 19] = "DISCOVER_FEED_THUMBNAIL";
    LegacyMushroomContentType[LegacyMushroomContentType["DISCOVER_PUBLISHER_SNAP"] = 20] = "DISCOVER_PUBLISHER_SNAP";
    LegacyMushroomContentType[LegacyMushroomContentType["DISCOVER_STORY_SNAP"] = 21] = "DISCOVER_STORY_SNAP";
    LegacyMushroomContentType[LegacyMushroomContentType["DISCOVER_STORY_STREAMING_SNAP"] = 22] = "DISCOVER_STORY_STREAMING_SNAP";
    LegacyMushroomContentType[LegacyMushroomContentType["EMOJI"] = 23] = "EMOJI";
    LegacyMushroomContentType[LegacyMushroomContentType["EMOJI_BRUSH"] = 24] = "EMOJI_BRUSH";
    LegacyMushroomContentType[LegacyMushroomContentType["EXTERNAL_GEOFILTER"] = 25] = "EXTERNAL_GEOFILTER";
    LegacyMushroomContentType[LegacyMushroomContentType["EXTERNAL_STICKER"] = 26] = "EXTERNAL_STICKER";
    LegacyMushroomContentType[LegacyMushroomContentType["GEOFILTER_OVERLAY"] = 27] = "GEOFILTER_OVERLAY";
    LegacyMushroomContentType[LegacyMushroomContentType["IMAGELOADING_URL"] = 28] = "IMAGELOADING_URL";
    LegacyMushroomContentType[LegacyMushroomContentType["HELVETICA"] = 29] = "HELVETICA";
    LegacyMushroomContentType[LegacyMushroomContentType["INFO_STICKER"] = 30] = "INFO_STICKER";
    LegacyMushroomContentType[LegacyMushroomContentType["LENS_CONTENT"] = 31] = "LENS_CONTENT";
    LegacyMushroomContentType[LegacyMushroomContentType["LENS_ICON"] = 32] = "LENS_ICON";
    LegacyMushroomContentType[LegacyMushroomContentType["LENS_OVERLAY_IMAGE"] = 33] = "LENS_OVERLAY_IMAGE";
    LegacyMushroomContentType[LegacyMushroomContentType["LENS_REMOTE_ASSET"] = 34] = "LENS_REMOTE_ASSET";
    LegacyMushroomContentType[LegacyMushroomContentType["LOGIN_KIT_PRIVACT"] = 35] = "LOGIN_KIT_PRIVACT";
    LegacyMushroomContentType[LegacyMushroomContentType["MAPS_EGGHUNT"] = 36] = "MAPS_EGGHUNT";
    LegacyMushroomContentType[LegacyMushroomContentType["MAPS_KASHMIR"] = 37] = "MAPS_KASHMIR";
    LegacyMushroomContentType[LegacyMushroomContentType["MAPS_WORLDEFFECTS"] = 38] = "MAPS_WORLDEFFECTS";
    LegacyMushroomContentType[LegacyMushroomContentType["MEDIA_PACKAGE_COMPOSITE"] = 39] = "MEDIA_PACKAGE_COMPOSITE";
    LegacyMushroomContentType[LegacyMushroomContentType["MEDIA_PACKAGE_THUMB"] = 40] = "MEDIA_PACKAGE_THUMB";
    LegacyMushroomContentType[LegacyMushroomContentType["MEMORIES_EDITS"] = 41] = "MEMORIES_EDITS";
    LegacyMushroomContentType[LegacyMushroomContentType["MEMORIES_MEDIA"] = 42] = "MEMORIES_MEDIA";
    LegacyMushroomContentType[LegacyMushroomContentType["MEMORIES_MINI_THUMBNAIL"] = 43] = "MEMORIES_MINI_THUMBNAIL";
    LegacyMushroomContentType[LegacyMushroomContentType["MEMORIES_OVERLAY"] = 44] = "MEMORIES_OVERLAY";
    LegacyMushroomContentType[LegacyMushroomContentType["MEMORIES_THUMBNAIL"] = 45] = "MEMORIES_THUMBNAIL";
    LegacyMushroomContentType[LegacyMushroomContentType["NON_USER_BITMOJI"] = 46] = "NON_USER_BITMOJI";
    LegacyMushroomContentType[LegacyMushroomContentType["PAYMENTS"] = 47] = "PAYMENTS";
    LegacyMushroomContentType[LegacyMushroomContentType["PUBLISHER_SNAP_MEDIA"] = 48] = "PUBLISHER_SNAP_MEDIA";
    LegacyMushroomContentType[LegacyMushroomContentType["SHAZAM"] = 49] = "SHAZAM";
    LegacyMushroomContentType[LegacyMushroomContentType["SKY_MODEL"] = 50] = "SKY_MODEL";
    LegacyMushroomContentType[LegacyMushroomContentType["SNAP"] = 51] = "SNAP";
    LegacyMushroomContentType[LegacyMushroomContentType["SNAP_FIRST_FRAME"] = 52] = "SNAP_FIRST_FRAME";
    LegacyMushroomContentType[LegacyMushroomContentType["SNAP_LOADING_FRAME"] = 53] = "SNAP_LOADING_FRAME";
    LegacyMushroomContentType[LegacyMushroomContentType["SNAP_STICKER"] = 54] = "SNAP_STICKER";
    LegacyMushroomContentType[LegacyMushroomContentType["SPECTACLES"] = 55] = "SPECTACLES";
    LegacyMushroomContentType[LegacyMushroomContentType["STICKER_TAG"] = 56] = "STICKER_TAG";
    LegacyMushroomContentType[LegacyMushroomContentType["STICKERS_METADATA"] = 57] = "STICKERS_METADATA";
    LegacyMushroomContentType[LegacyMushroomContentType["STORY_SNAP"] = 58] = "STORY_SNAP";
    LegacyMushroomContentType[LegacyMushroomContentType["STORY_THUMB"] = 59] = "STORY_THUMB";
    LegacyMushroomContentType[LegacyMushroomContentType["TRACKING_DATA"] = 60] = "TRACKING_DATA";
    LegacyMushroomContentType[LegacyMushroomContentType["USER_GENERATED_ASSETS"] = 61] = "USER_GENERATED_ASSETS";
    LegacyMushroomContentType[LegacyMushroomContentType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(LegacyMushroomContentType || (LegacyMushroomContentType = {}));
function createBaseContentManagerSignals() {
    return { legacyMushroomContentType: 0 };
}
export const ContentManagerSignals = {
    encode(message, writer = new BinaryWriter()) {
        if (message.legacyMushroomContentType !== 0) {
            writer.uint32(8).int32(message.legacyMushroomContentType);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseContentManagerSignals();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.legacyMushroomContentType = reader.int32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    create(base) {
        return ContentManagerSignals.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseContentManagerSignals();
        message.legacyMushroomContentType = (_a = object.legacyMushroomContentType) !== null && _a !== void 0 ? _a : 0;
        return message;
    },
};
