// @snap/camera-kit/dist/generated-proto/pb_schema/cdp/cof/creative_tools_provided_signals.js
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
export const protobufPackage = "snapchat.cdp.cof";
function createBaseCreativeToolsSignals() {
    return { withAnimatedOverlay: false, withMusic: false };
}
export const CreativeToolsSignals = {
    encode(message, writer = new BinaryWriter()) {
        if (message.withAnimatedOverlay !== false) {
            writer.uint32(8).bool(message.withAnimatedOverlay);
        }
        if (message.withMusic !== false) {
            writer.uint32(16).bool(message.withMusic);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCreativeToolsSignals();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.withAnimatedOverlay = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.withMusic = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    create(base) {
        return CreativeToolsSignals.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseCreativeToolsSignals();
        message.withAnimatedOverlay = (_a = object.withAnimatedOverlay) !== null && _a !== void 0 ? _a : false;
        message.withMusic = (_b = object.withMusic) !== null && _b !== void 0 ? _b : false;
        return message;
    },
};
