// @snap/camera-kit/dist/generated-proto/pb_schema/cdp/cof/lenses_provided_signals.js
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
export const protobufPackage = "snapchat.cdp.cof";
function createBaseLensesSignals() {
    return { lastLensActivatedTime: "0", lastSnappableLensActivatedTime: "0", isAnyEffectApplied: false };
}
export const LensesSignals = {
    encode(message, writer = new BinaryWriter()) {
        if (message.lastLensActivatedTime !== "0") {
            writer.uint32(8).uint64(message.lastLensActivatedTime);
        }
        if (message.lastSnappableLensActivatedTime !== "0") {
            writer.uint32(16).uint64(message.lastSnappableLensActivatedTime);
        }
        if (message.isAnyEffectApplied !== false) {
            writer.uint32(24).bool(message.isAnyEffectApplied);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseLensesSignals();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.lastLensActivatedTime = reader.uint64().toString();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.lastSnappableLensActivatedTime = reader.uint64().toString();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.isAnyEffectApplied = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    create(base) {
        return LensesSignals.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseLensesSignals();
        message.lastLensActivatedTime = (_a = object.lastLensActivatedTime) !== null && _a !== void 0 ? _a : "0";
        message.lastSnappableLensActivatedTime = (_b = object.lastSnappableLensActivatedTime) !== null && _b !== void 0 ? _b : "0";
        message.isAnyEffectApplied = (_c = object.isAnyEffectApplied) !== null && _c !== void 0 ? _c : false;
        return message;
    },
};
