// @snap/camera-kit/dist/generated-proto/pb_schema/cdp/cof/opera_provided_signals.js
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
export const protobufPackage = "snapchat.cdp.cof";
function createBaseOperaSignals() {
    return { contentViewSource: 0, playbackItemType: 0, useVerticalNavigation: false };
}
export const OperaSignals = {
    encode(message, writer = new BinaryWriter()) {
        if (message.contentViewSource !== 0) {
            writer.uint32(8).int32(message.contentViewSource);
        }
        if (message.playbackItemType !== 0) {
            writer.uint32(16).int32(message.playbackItemType);
        }
        if (message.useVerticalNavigation !== false) {
            writer.uint32(24).bool(message.useVerticalNavigation);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOperaSignals();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.contentViewSource = reader.int32();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.playbackItemType = reader.int32();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.useVerticalNavigation = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    create(base) {
        return OperaSignals.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseOperaSignals();
        message.contentViewSource = (_a = object.contentViewSource) !== null && _a !== void 0 ? _a : 0;
        message.playbackItemType = (_b = object.playbackItemType) !== null && _b !== void 0 ? _b : 0;
        message.useVerticalNavigation = (_c = object.useVerticalNavigation) !== null && _c !== void 0 ? _c : false;
        return message;
    },
};
