// @snap/camera-kit/dist/generated-proto/pb_schema/snapdoc/extensions.js
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
export const protobufPackage = "snapchat.snapdoc";
function createBaseSnapDocMessageOptions() {
    return { componentTypeId: undefined, componentTypeName: undefined, shouldEncrypt: false };
}
export const SnapDocMessageOptions = {
    encode(message, writer = new BinaryWriter()) {
        if (message.componentTypeId !== undefined) {
            writer.uint32(8).int32(message.componentTypeId);
        }
        if (message.componentTypeName !== undefined) {
            writer.uint32(18).string(message.componentTypeName);
        }
        if (message.shouldEncrypt !== false) {
            writer.uint32(24).bool(message.shouldEncrypt);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSnapDocMessageOptions();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.componentTypeId = reader.int32();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.componentTypeName = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.shouldEncrypt = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    create(base) {
        return SnapDocMessageOptions.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseSnapDocMessageOptions();
        message.componentTypeId = (_a = object.componentTypeId) !== null && _a !== void 0 ? _a : undefined;
        message.componentTypeName = (_b = object.componentTypeName) !== null && _b !== void 0 ? _b : undefined;
        message.shouldEncrypt = (_c = object.shouldEncrypt) !== null && _c !== void 0 ? _c : false;
        return message;
    },
};
