// @snap/camera-kit/dist/generated-proto/pb_schema/snapdoc/media_codec_formats.js
export const protobufPackage = "snapchat.snapdoc";
export var AudioCodecFormat;
(function (AudioCodecFormat) {
    AudioCodecFormat[AudioCodecFormat["AUDIO_UNSET"] = 0] = "AUDIO_UNSET";
    AudioCodecFormat[AudioCodecFormat["AAC"] = 1] = "AAC";
    AudioCodecFormat[AudioCodecFormat["HEAAC"] = 2] = "HEAAC";
    AudioCodecFormat[AudioCodecFormat["OPUS"] = 3] = "OPUS";
    AudioCodecFormat[AudioCodecFormat["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(AudioCodecFormat || (AudioCodecFormat = {}));
export var ImageCodecFormat;
(function (ImageCodecFormat) {
    ImageCodecFormat[ImageCodecFormat["IMAGE_UNSET"] = 0] = "IMAGE_UNSET";
    ImageCodecFormat[ImageCodecFormat["JPEG"] = 1] = "JPEG";
    ImageCodecFormat[ImageCodecFormat["PNG"] = 2] = "PNG";
    ImageCodecFormat[ImageCodecFormat["WEBP"] = 3] = "WEBP";
    ImageCodecFormat[ImageCodecFormat["HEIF"] = 4] = "HEIF";
    ImageCodecFormat[ImageCodecFormat["AWEBP"] = 5] = "AWEBP";
    ImageCodecFormat[ImageCodecFormat["GIF"] = 6] = "GIF";
    ImageCodecFormat[ImageCodecFormat["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(ImageCodecFormat || (ImageCodecFormat = {}));
export var VideoCodecFormat;
(function (VideoCodecFormat) {
    VideoCodecFormat[VideoCodecFormat["VIDEO_UNSET"] = 0] = "VIDEO_UNSET";
    VideoCodecFormat[VideoCodecFormat["H264"] = 1] = "H264";
    VideoCodecFormat[VideoCodecFormat["HEVC"] = 2] = "HEVC";
    VideoCodecFormat[VideoCodecFormat["AV1"] = 3] = "AV1";
    VideoCodecFormat[VideoCodecFormat["VP9"] = 4] = "VP9";
    VideoCodecFormat[VideoCodecFormat["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(VideoCodecFormat || (VideoCodecFormat = {}));
