// @snap/camera-kit/dist/generated-proto/pb_schema/snapdoc/media_id.js
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
export const protobufPackage = "snapchat.snapdoc";
function createBaseMediaId() {
    return { mediaListId: "0" };
}
export const MediaId = {
    encode(message, writer = new BinaryWriter()) {
        if (message.mediaListId !== "0") {
            writer.uint32(8).int64(message.mediaListId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMediaId();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.mediaListId = reader.int64().toString();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    create(base) {
        return MediaId.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseMediaId();
        message.mediaListId = (_a = object.mediaListId) !== null && _a !== void 0 ? _a : "0";
        return message;
    },
};
