// @snap/camera-kit/dist/generated-proto/pb_schema/snapdoc/snap_component_type.js
export const protobufPackage = "snapchat.snapdoc";
export var SnapComponentType;
(function (SnapComponentType) {
    SnapComponentType[SnapComponentType["UNKNOWN"] = 0] = "UNKNOWN";
    SnapComponentType[SnapComponentType["MEDIA"] = 1] = "MEDIA";
    SnapComponentType[SnapComponentType["DURATION"] = 2] = "DURATION";
    SnapComponentType[SnapComponentType["TIMING"] = 3] = "TIMING";
    SnapComponentType[SnapComponentType["UNLOCKABLES"] = 4] = "UNLOCKABLES";
    SnapComponentType[SnapComponentType["CONTEXT"] = 5] = "CONTEXT";
    SnapComponentType[SnapComponentType["ATTACHMENT"] = 6] = "ATTACHMENT";
    SnapComponentType[SnapComponentType["SNAPPABLE"] = 7] = "SNAPPABLE";
    SnapComponentType[SnapComponentType["STORY_METADATA"] = 8] = "STORY_METADATA";
    SnapComponentType[SnapComponentType["GROUP_STORY"] = 23] = "GROUP_STORY";
    SnapComponentType[SnapComponentType["OUR_STORY"] = 24] = "OUR_STORY";
    SnapComponentType[SnapComponentType["ATTRIBUTION"] = 9] = "ATTRIBUTION";
    SnapComponentType[SnapComponentType["THUMBNAIL"] = 10] = "THUMBNAIL";
    SnapComponentType[SnapComponentType["PREVIEW_FRAME"] = 46] = "PREVIEW_FRAME";
    SnapComponentType[SnapComponentType["CONTENT_ANALYSIS"] = 11] = "CONTENT_ANALYSIS";
    SnapComponentType[SnapComponentType["PROVENANCE"] = 22] = "PROVENANCE";
    SnapComponentType[SnapComponentType["PACKAGING"] = 25] = "PACKAGING";
    SnapComponentType[SnapComponentType["CONTENT_TAGGING"] = 26] = "CONTENT_TAGGING";
    SnapComponentType[SnapComponentType["SNAP_MODERATION"] = 30] = "SNAP_MODERATION";
    SnapComponentType[SnapComponentType["CREATIVE_KIT_SOURCE_APP"] = 31] = "CREATIVE_KIT_SOURCE_APP";
    SnapComponentType[SnapComponentType["CONTENT_OBJECTS"] = 32] = "CONTENT_OBJECTS";
    SnapComponentType[SnapComponentType["LEGACY_MULTISNAP"] = 33] = "LEGACY_MULTISNAP";
    SnapComponentType[SnapComponentType["SPECTACLES"] = 34] = "SPECTACLES";
    SnapComponentType[SnapComponentType["CAPTION"] = 35] = "CAPTION";
    SnapComponentType[SnapComponentType["VIDEO_TRANSCRIPTION"] = 36] = "VIDEO_TRANSCRIPTION";
    SnapComponentType[SnapComponentType["REPLY"] = 37] = "REPLY";
    SnapComponentType[SnapComponentType["GAME_ATTACHMENT"] = 38] = "GAME_ATTACHMENT";
    SnapComponentType[SnapComponentType["VIDEO_INTERVAL"] = 39] = "VIDEO_INTERVAL";
    SnapComponentType[SnapComponentType["PERMITTED_USER_ACTIONS"] = 40] = "PERMITTED_USER_ACTIONS";
    SnapComponentType[SnapComponentType["CREATIVE_TOOLS_ANALYTICS"] = 41] = "CREATIVE_TOOLS_ANALYTICS";
    SnapComponentType[SnapComponentType["PLAYBACK_CHARACTERISTICS"] = 42] = "PLAYBACK_CHARACTERISTICS";
    SnapComponentType[SnapComponentType["CAPTURE_CHARACTERISTICS"] = 43] = "CAPTURE_CHARACTERISTICS";
    SnapComponentType[SnapComponentType["STICKER_METADATA"] = 44] = "STICKER_METADATA";
    SnapComponentType[SnapComponentType["LENS"] = 45] = "LENS";
    SnapComponentType[SnapComponentType["ATTACHMENTS"] = 47] = "ATTACHMENTS";
    SnapComponentType[SnapComponentType["GENERATED_CAMEO"] = 49] = "GENERATED_CAMEO";
    SnapComponentType[SnapComponentType["SNAP_PUBLISHER_CREATION_METADATA"] = 50] = "SNAP_PUBLISHER_CREATION_METADATA";
    SnapComponentType[SnapComponentType["DISTRIBUTION"] = 51] = "DISTRIBUTION";
    SnapComponentType[SnapComponentType["SUBTITLES"] = 52] = "SUBTITLES";
    SnapComponentType[SnapComponentType["CAMEO_SNAP_RESOURCE"] = 53] = "CAMEO_SNAP_RESOURCE";
    SnapComponentType[SnapComponentType["CAMEO_METADATA"] = 54] = "CAMEO_METADATA";
    SnapComponentType[SnapComponentType["LOCATION"] = 55] = "LOCATION";
    SnapComponentType[SnapComponentType["CALL_TO_ACTION"] = 48] = "CALL_TO_ACTION";
    SnapComponentType[SnapComponentType["APP_DEEP_LINK"] = 12] = "APP_DEEP_LINK";
    SnapComponentType[SnapComponentType["APP_INSTALL"] = 13] = "APP_INSTALL";
    SnapComponentType[SnapComponentType["COMMERCE"] = 14] = "COMMERCE";
    SnapComponentType[SnapComponentType["LENS_CAROUSEL"] = 15] = "LENS_CAROUSEL";
    SnapComponentType[SnapComponentType["LONGFORM_VIDEO"] = 16] = "LONGFORM_VIDEO";
    SnapComponentType[SnapComponentType["NOTIFICATION_SETTINGS"] = 17] = "NOTIFICATION_SETTINGS";
    SnapComponentType[SnapComponentType["SUBSCRIPTION"] = 18] = "SUBSCRIPTION";
    SnapComponentType[SnapComponentType["WEB_PAGE"] = 19] = "WEB_PAGE";
    SnapComponentType[SnapComponentType["ARTICLE"] = 27] = "ARTICLE";
    SnapComponentType[SnapComponentType["POLL"] = 28] = "POLL";
    SnapComponentType[SnapComponentType["CAMERA_ATTACHMENT"] = 29] = "CAMERA_ATTACHMENT";
    SnapComponentType[SnapComponentType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(SnapComponentType || (SnapComponentType = {}));
